File: python/glide-async/python/glide/isolated_scope.py
# Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

"""
Async Isolated Execution Scope for the GLIDE async client (Feature 2).

Provides a dedicated connection for operations requiring per-connection state:
WATCH/MULTI/EXEC, CLIENT TRACKING, blocking commands. Commands bypass the
multiplexer and execute on a single TCP connection.

Usage:
    from glide import GlideClient
    from glide.config import GlideClientConfiguration, NodeAddress

    client = await GlideClient.create(GlideClientConfiguration([NodeAddress("localhost", 6379)]))

    async with client.scoped_connection() as scope:
        await scope.watch("counter")
        val = await scope.get("counter")
        await scope.multi()
        await scope.set("counter", str(int(val) + 1))
        result = await scope.exec()
"""

import asyncio
import struct
from typing import Optional

from glide_shared._glide_ffi import _GlideFFI


class AsyncIsolatedScope:
    """
    Async wrapper around a dedicated scoped connection (Feature 2).

    Commands are dispatched to a thread pool executor (since the underlying
    FFI call blocks on Tokio), keeping the asyncio event loop free.

    Thread Safety: A single AsyncIsolatedScope should only be used from one
    coroutine at a time. Multiple coroutines should each acquire their own scope.

    Use as an async context manager:
        async with client.scoped_connection() as scope:
            await scope.watch("key")
            ...
    """

    __slots__ = (
        "_scope_id",
        "_client_id",
        "_ffi",
        "_lib",
        "_released",
        "_response_parser",
        "_executor",
    )

    def __init__(
        self,
        scope_id: int,
        client_id: int,
        ffi_instance: _GlideFFI,
        response_parser,
        executor=None,
    ):
        self._scope_id = scope_id
        self._client_id = client_id
        self._ffi = ffi_instance.ffi
        self._lib = ffi_instance.lib
        self._released = False
        self._response_parser = response_parser
        self._executor = executor  # ThreadPoolExecutor or None (uses default)

    @property
    def scope_id(self) -> int:
        if self._released:
            raise RuntimeError("Scope already released")
        return self._scope_id

    @property
    def is_released(self) -> bool:
        return self._released

    # ═══════════════════════════════════════════════════════════════════════════
    # COMMANDS (all async)
    # ═══════════════════════════════════════════════════════════════════════════

    async def watch(self, *keys: str) -> Optional[str]:
        """WATCH one or more keys for optimistic locking."""
        return await self._cmd("WATCH", *keys)

    async def unwatch(self) -> Optional[str]:
        """Discard all watched keys."""
        return await self._cmd("UNWATCH")

    async def multi(self) -> Optional[str]:
        """Begin a transaction block."""
        return await self._cmd("MULTI")

    async def exec(self) -> Optional[str]:
        """Execute the transaction (returns None if WATCH detected a conflict)."""
        return await self._cmd("EXEC")

    async def discard(self) -> Optional[str]:
        """Discard the transaction."""
        return await self._cmd("DISCARD")

    async def get(self, key: str) -> Optional[str]:
        """GET a key value."""
        return await self._cmd("GET", key)

    async def set(self, key: str, value: str) -> Optional[str]:
        """SET a key to a value."""
        return await self._cmd("SET", key, value)

    async def incr(self, key: str) -> Optional[str]:
        """Increment a key's integer value by 1."""
        return await self._cmd("INCR", key)

    async def ping(self, message: Optional[str] = None) -> Optional[str]:
        """PING the server."""
        if message:
            return await self._cmd("PING", message)
        return await self._cmd("PING")

    async def select(self, db: int) -> Optional[str]:
        """SELECT a database by index."""
        return await self._cmd("SELECT", str(db))

    async def execute_command(self, command: str, *args: str) -> Optional[str]:
        """Execute an arbitrary command on this scope."""
        return await self._cmd(command, *args)

    # ═══════════════════════════════════════════════════════════════════════════
    # INTERNAL
    # ═══════════════════════════════════════════════════════════════════════════

    async def _cmd(self, command: str, *args: str) -> Optional[str]:
        """Serialize and execute a command on the scoped connection (async)."""
        if self._released:
            raise RuntimeError("Scope already released")

        # Serialize in wire format
        cmd_bytes = command.encode("utf-8")
        parts = [
            struct.pack("<I", len(cmd_bytes)),
            cmd_bytes,
            struct.pack("<I", len(args)),
        ]
        for arg in args:
            arg_bytes = arg.encode("utf-8")
            parts.append(struct.pack("<I", len(arg_bytes)))
            parts.append(arg_bytes)
        payload = b"".join(parts)

        # Execute in thread pool to avoid blocking the event loop
        loop = asyncio.get_running_loop()
        result_ptr = await loop.run_in_executor(
            self._executor, self._execute_sync, payload
        )

        if result_ptr == self._ffi.NULL:
            raise RuntimeError(f"Scope execute failed: invalid scope {self._scope_id}")

        try:
            return self._parse_result(result_ptr)
        finally:
            self._lib.free_command_result(result_ptr)

    def _execute_sync(self, payload: bytes):
        """Synchronous FFI call (runs in executor thread)."""
        buf = self._ffi.from_buffer(payload)
        return self._lib.glide_scope_execute(
            self._scope_id,
            self._ffi.cast("const uint8_t*", buf),
            len(payload),
        )

    def _parse_result(self, result_ptr) -> Optional[str]:
        """Parse a CommandResult pointer into a Python value."""
        result = result_ptr

        if result.command_error != self._ffi.NULL:
            err = result.command_error
            msg = self._ffi.string(err.command_error_message).decode("utf-8")
            raise RuntimeError(f"Scope command error: {msg}")

        if result.response == self._ffi.NULL:
            return None

        return self._response_parser(result.response)

    # ═══════════════════════════════════════════════════════════════════════════
    # ASYNC CONTEXT MANAGER
    # ═══════════════════════════════════════════════════════════════════════════

    async def close(self):
        """Release the scope back to the pool."""
        if not self._released:
            self._released = True
            self._lib.glide_scope_release(self._scope_id, self._client_id)

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()
        return False

    def __del__(self):
        """Safety net — release on GC if not explicitly closed."""
        if not self._released:
            try:
                self._released = True
                self._lib.glide_scope_release(self._scope_id, self._client_id)
            except Exception:
                pass

File: node/src/IsolatedScope.ts
/**
 * Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0
 *
 * Isolated Execution Scope for Node.js.
 *
 * Commands go through glide-core's scope module via N-API:
 * - `scopeTryAcquire` → `glide_core::scope::try_acquire_scope`
 * - `scopeExecute` → `glide_core::scope::execute_scope_command`
 * - `scopeRelease` → `glide_core::scope::release_scope`
 *
 * This gives us state tracking, cluster slot pinning, compression,
 * timeout, and zero-cost release — identical to Java/Python/Go.
 */

import { scopeExecute, scopeRelease } from "../build-ts/native";
import type { GlideString } from "./BaseClient";
import type { BaseClient } from "./BaseClient";
import { hasScopeConnectionRequest, tryAcquireScope } from "./ScopeInternal.js";

// ─── Wire Format Serialization ───────────────────────────────────────────────

/**
 * Serialize a command into the wire format expected by glide-core::scope::deserialize_command.
 *
 * Wire format (little-endian):
 *   [4:cmd_name_len][cmd_name][4:num_args][4:arg1_len][arg1]...[4:argN_len][argN]
 */
function serializeCommand(cmdName: string, args: GlideString[]): Uint8Array {
    const encoder = new TextEncoder();
    const cmdBytes = encoder.encode(cmdName);

    let totalSize = 4 + cmdBytes.length + 4;
    const argBuffers: Uint8Array[] = [];

    for (const arg of args) {
        const buf =
            arg instanceof Buffer
                ? new Uint8Array(arg)
                : encoder.encode(arg as string);
        argBuffers.push(buf);
        totalSize += 4 + buf.length;
    }

    const buffer = new Uint8Array(totalSize);
    const view = new DataView(buffer.buffer);
    let offset = 0;

    view.setUint32(offset, cmdBytes.length, true);
    offset += 4;
    buffer.set(cmdBytes, offset);
    offset += cmdBytes.length;

    view.setUint32(offset, argBuffers.length, true);
    offset += 4;

    for (const argBuf of argBuffers) {
        view.setUint32(offset, argBuf.length, true);
        offset += 4;
        buffer.set(argBuf, offset);
        offset += argBuf.length;
    }

    return buffer;
}

// ─── Slot Computation ────────────────────────────────────────────────────────

/**
 * Compute the Redis cluster hash slot for a key (CRC16 mod 16384).
 * Handles hash tags: if the key contains {...}, only the content between
 * the first { and first } is hashed.
 */
function slotForKey(key: Buffer): number {
    const start = key.indexOf(0x7b); // '{'

    if (start !== -1) {
        const end = key.indexOf(0x7d, start + 1); // '}'

        if (end !== -1 && end !== start + 1) {
            key = key.subarray(start + 1, end);
        }
    }

    let crc = 0;

    for (const byte of key) {
        crc ^= byte << 8;

        for (let j = 0; j < 8; j++) {
            if (crc & 0x8000) {
                crc = ((crc << 1) ^ 0x1021) & 0xffff;
            } else {
                crc = (crc << 1) & 0xffff;
            }
        }
    }

    return crc % 16384;
}

// ─── IsolatedScope ───────────────────────────────────────────────────────────

/**
 * An isolated execution scope backed by glide-core's scope module.
 *
 * Each scope has a dedicated TCP connection managed by the Rust core.
 * Commands go through `execute_scope_command` which provides:
 * - Per-connection state tracking (WATCH, MULTI, SELECT, subscriptions)
 * - Cluster slot pinning (first keyed command pins the slot)
 * - Cross-slot rejection in cluster mode
 * - Compression/decompression using parent client settings
 * - Timeout handling via parent client's request_timeout
 * - Zero-cost release when state is clean
 *
 * @example
 * ```typescript
 * const scope = await IsolatedScope.acquire(client);
 * try {
 *     await scope.watch("key");
 *     const val = await scope.get("key");
 *     await scope.multi();
 *     await scope.set("key", "new_value");
 *     const result = await scope.exec();
 * } finally {
 *     scope.release();
 * }
 * ```
 */
export class IsolatedScope {
    private scopeId: number;
    private clientId: number;
    private released = false;

    private constructor(scopeId: number, clientId: number) {
        this.scopeId = scopeId;
        this.clientId = clientId;
    }

    /**
     * Acquire an isolated scope for a client.
     *
     * Uses `glide_core::scope::try_acquire_scope` with exponential backoff.
     * Creates new scope connections in the background if the pool is empty.
     *
     * The connection request is taken from the client (captured when it
     * connected), so the one-argument form `IsolatedScope.acquire(client)`
     * works on both a standalone client and a pool-borrowed client. Passing
     * `connectionRequestBytes` explicitly is still supported and overrides the
     * client's captured request.
     *
     * Prefer {@link BaseClient.scopedConnection} for the common case.
     *
     * @param client - A GlideClient or GlideClusterClient instance.
     * @param routingKey - In cluster mode, the key whose hash slot determines which node the scope connects to.
     * @param maxRetries - Maximum retries with backoff. Default: 10.
     * @returns A new IsolatedScope.
     */
    static async acquire(
        client: BaseClient,
        routingKey?: string,
        maxRetries?: number,
    ): Promise<IsolatedScope>;
    /**
     * Acquire an isolated scope for a client using an explicit connection request.
     *
     * @param client - A GlideClient or GlideClusterClient instance.
     * @param connectionRequestBytes - Serialized connection request to use for the scope's connection.
     * @param routingKey - In cluster mode, the key whose hash slot determines which node the scope connects to.
     * @param maxRetries - Maximum retries with backoff. Default: 10.
     * @returns A new IsolatedScope.
     */
    static async acquire(
        client: BaseClient,
        connectionRequestBytes: Uint8Array,
        routingKey?: string,
        maxRetries?: number,
    ): Promise<IsolatedScope>;
    static async acquire(
        client: BaseClient,
        bytesOrRoutingKey?: Uint8Array | string,
        routingKeyOrMaxRetries?: string | number,
        maxRetries = 10,
    ): Promise<IsolatedScope> {
        // Normalize the overloaded arguments. When the second argument is a
        // Uint8Array it is an explicit connection request; otherwise it is the
        // routing key and the bytes are derived from the client.
        let explicitBytes: Uint8Array | undefined;
        let routingKey: string | undefined;

        if (bytesOrRoutingKey instanceof Uint8Array) {
            // acquire(client, bytes, routingKey?, maxRetries?)
            explicitBytes = bytesOrRoutingKey;
            routingKey = routingKeyOrMaxRetries as string | undefined;
        } else {
            // acquire(client, routingKey?, maxRetries?)
            routingKey = bytesOrRoutingKey;

            if (typeof routingKeyOrMaxRetries === "number") {
                maxRetries = routingKeyOrMaxRetries;
            }
        }

        // Without explicit bytes, the scope reuses the client's cached request.
        // It lives in module-private storage, so acquisition happens through a
        // helper that never hands the bytes back here.
        if (!explicitBytes && !hasScopeConnectionRequest(client)) {
            throw new Error(
                "Client has no connection request available for a scope. " +
                    "Ensure it was created via GlideClient.createClient() (or GlideClusterClient.createClient()) and is connected.",
            );
        }

        // Get the client_id from the handle (registered in Rust scope registry)
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const clientId = (client as any).clientHandle?.clientId;

        if (clientId === undefined || clientId === null) {
            throw new Error(
                "Client does not have a valid handle. Ensure it was created via GlideClient.createClient().",
            );
        }

        const routingSlot = routingKey
            ? slotForKey(Buffer.from(routingKey))
            : 0;

        let backoffMs = 10;

        for (let i = 0; i < maxRetries; i++) {
            const scopeId = tryAcquireScope(
                client,
                clientId,
                routingSlot,
                explicitBytes,
            );

            if (scopeId >= 0) {
                return new IsolatedScope(scopeId, clientId);
            }

            // Exponential backoff
            await new Promise((resolve) => setTimeout(resolve, backoffMs));
            backoffMs = Math.min(backoffMs * 2, 500);
        }

        throw new Error(
            "Could not acquire scoped connection after retries. " +
                "All scope connections may be in use or connection creation failed.",
        );
    }

    /** Whether this scope has been released. */
    get isReleased(): boolean {
        return this.released;
    }

    /**
     * Execute a command on this scope via glide-core.
     *
     * @param command - Command name (e.g., "GET", "SET", "WATCH").
     * @param args - Command arguments.
     * @returns Result as string, or null for nil.
     */
    async executeCommand(
        command: string,
        ...args: GlideString[]
    ): Promise<string | null> {
        if (this.released) {
            throw new Error("Scope has been released");
        }

        const cmdBytes = serializeCommand(command, args);
        return scopeExecute(this.scopeId, this.clientId, cmdBytes);
    }

    // ─── WATCH/MULTI/EXEC ────────────────────────────────────────────────────

    /** WATCH one or more keys for optimistic locking. */
    async watch(...keys: string[]): Promise<string | null> {
        return this.executeCommand("WATCH", ...keys);
    }

    /** UNWATCH all watched keys. */
    async unwatch(): Promise<string | null> {
        return this.executeCommand("UNWATCH");
    }

    /** Begin a MULTI transaction block. */
    async multi(): Promise<string | null> {
        return this.executeCommand("MULTI");
    }

    /**
     * Execute the MULTI transaction.
     * Returns null if WATCH detected a conflict (transaction aborted).
     */
    async exec(): Promise<string | null> {
        return this.executeCommand("EXEC");
    }

    /** DISCARD the current transaction. */
    async discard(): Promise<string | null> {
        return this.executeCommand("DISCARD");
    }

    // ─── Data Commands ───────────────────────────────────────────────────────

    /** GET a key's value. */
    async get(key: string): Promise<string | null> {
        return this.executeCommand("GET", key);
    }

    /** SET a key to a value. */
    async set(key: string, value: string): Promise<string | null> {
        return this.executeCommand("SET", key, value);
    }

    /** INCREMENT a key's integer value by 1. */
    async incr(key: string): Promise<string | null> {
        return this.executeCommand("INCR", key);
    }

    /** DEL one or more keys. */
    async del(...keys: string[]): Promise<string | null> {
        return this.executeCommand("DEL", ...keys);
    }

    // ─── Server Commands ─────────────────────────────────────────────────────

    /** PING the server. */
    async ping(): Promise<string | null> {
        return this.executeCommand("PING");
    }

    /** SELECT a database by index. */
    async select(db: number): Promise<string | null> {
        return this.executeCommand("SELECT", db.toString());
    }

    // ─── Lifecycle ───────────────────────────────────────────────────────────

    /**
     * Release this scope back to the connection pool.
     *
     * Goes through `glide_core::scope::release_scope` which handles:
     * - Zero-cost release if state is clean
     * - Async cleanup (DISCARD + UNWATCH + SELECT) if dirty
     * - Connection returned to pool for reuse
     *
     * Safe to call multiple times (idempotent).
     */
    release(): void {
        if (!this.released) {
            this.released = true;
            scopeRelease(this.scopeId, this.clientId);
        }
    }

    /** Alias for release(). */
    close(): void {
        this.release();
    }
}
